PILAR ARIAS, YEDRA MARTÍN, PATRICIA MEDIAVILLA, AMAYA SARASA

<GRUPO 9>

#Práctica 1

##Introducción

Esta es la primera práctica del segundo laboratorio. Dada una base de datos en formato .csv, se nos pide contestar a una serie de cuestiones. Para ello, lo priemro que debemos hacer es cargar y leer la base de datos:

In [3]:
!pip uninstall mlxtend
!pip install mlxtend



Found existing installation: mlxtend 0.21.0
Uninstalling mlxtend-0.21.0:
  Would remove:
    /usr/local/lib/python3.8/dist-packages/mlxtend-0.21.0.dist-info/*
    /usr/local/lib/python3.8/dist-packages/mlxtend/*
Proceed (Y/n)? y
  Successfully uninstalled mlxtend-0.21.0
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Using cached mlxtend-0.21.0-py2.py3-none-any.whl (1.3 MB)


In [1]:
#Importar librerías
#!apt install mlxtend
import numpy as np #numpy
import pandas as pd #pandas
import sklearn as sk #scikit-learn
import matplotlib.pyplot as plt #gráfica Matplotlib
from mlxtend.frequent_patterns import apriori
#from apyori import apriori #apriori
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import association_rules
#from mlxtend.frequent_patterns import frequent_patterns
#from mlxtend import frequent_patterns
#from mlxtend.frequent_patterns import frequent_patterns

#Importamos la base de datos
df = pd.read_csv('BlackFriday.csv',sep=',')
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 233599 entries, 0 to 233598
Data columns (total 11 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   User_ID                     233599 non-null  int64  
 1   Product_ID                  233599 non-null  object 
 2   Gender                      233599 non-null  object 
 3   Age                         233599 non-null  object 
 4   Occupation                  233599 non-null  int64  
 5   City_Category               233599 non-null  object 
 6   Stay_In_Current_City_Years  233599 non-null  object 
 7   Marital_Status              233599 non-null  int64  
 8   Product_Category_1          233599 non-null  int64  
 9   Product_Category_2          161255 non-null  float64
 10  Product_Category_3          71037 non-null   float64
dtypes: float64(2), int64(4), object(5)
memory usage: 19.6+ MB


User_ID Product_ID Gender    Age  Occupation City_Category  \
0  1000004  P00128942      M  46-50           7             B   
1  1000009  P00113442      M  26-35          17             C   
2  1000010  P00288442      F  36-45           1             B   
3  1000010  P00145342      F  36-45           1             B   
4  1000011  P00053842      F  26-35           1             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               1                   1   
1                          0               0                   3   
2                         4+               1                   5   
3                         4+               1                   4   
4                          1               0                   4   

   Product_Category_2  Product_Category_3  
0                11.0                 NaN  
1                 5.0                 NaN  
2                14.0                 NaN  
3                 9.0                 NaN  
4                 5.0                12.0

In [3]:
data1=df[['Product_ID']]

In [9]:
data1.value_counts()

Product_ID
P00265242     829
P00112142     717
P00025442     695
P00110742     680
P00046742     646
             ... 
P00291642       1
P00159042       1
P00352042       1
P00161042       1
P00161642       1
Length: 3491, dtype: int64

In [17]:
pilar=pd.get_dummies(df.Product_ID)

In [19]:
patricia=  pd.concat([[df['User_ID'], pilar] axis=0)

SyntaxError: ignored

In [ ]:
pilar['caca'] = pd.concat([df['User_ID'], pd.get_dummies(df.Product_ID)], axis=0)

In [2]:
df.set_index(pd.get_dummies(df.Product_ID), inplace=True)

TypeError: ignored

Subimos la base de datos, df.info nos aporta la infomación que contiene la base de datos, vemos el tipo, la cantidad de las variables con las que trabajamos... y .head vamos a ver que forma tiene la base de datos y las primeras filas de esta.

Según se ha estudiado en clase, encontramos dos métodos para afrontarlo. Nos decantaremos por CRISP-DM. Las siglas para: CRoss-Industry Standard Process for Data Mining. 

Es un modelo de proceso de minería de datos que se utiliza en la industria. Es un marco de trabajo que define un proceso estandarizado para llevar a cabo proyectos de minería de datos, desde la definición del problema hasta la implementación y evaluación del resultado. Los seis pasos que componen Crisp-DM son los siguientes:

- Comprensión de los negocios y definición del problema: comprender la naturaleza del problema y los objetivos de negocio.

- Comprensión de los datos: recopilación inicial, descripción y exploración  de datos.

- Preparación de los datos: Selección, limpieza, construcción, integración, formateo de los datos.

- Modelado: aplicar las técnicas de minería de datos a los dataset.

- Evaluación: determinar si los resultados son útiles a las necesidades del negocio.

- Despliegue: explotar utilidad de los modelos, integrándolos en las tareas de toma de decisiones de la organización -> Call to Action. 

##Comprensión de los negocios y definición del problema

La librería se usará de la siguiente manera:
1) Empezaremos obteniendo los itemsets frecuentes para k=1. En este punto necesitaremos obtener el
soporte de los itemset. Por lo tanto, se tendrá que usar un método que dado un itemset devuelva su
soporte.

2) Para k≥2.
a. Mostrar los itemset frecuentes candidatos y su soporte.

3) Repetir el proceso 2 hasta que no se generen nuevos itemsets frecuentes.

4) Mostrar todas las posibles reglas con la confianza de cada una de ellas.

5) Listar todas las reglas que sean de alta confianza.

6) Usar los siguientes métodos. Dado un antecesor, devolver todas las reglas que contengan a dicho
antecesor. Dado un umbral mínimo devolver todas las reglas que cumplan con dicha confianza.

7) Utiliza las representaciones gráficas que consideres adecuadas para representar las reglas obtenidas
y obtener conclusiones a partir de los datos.

A partir de lo anterior de pide:
- Prueba al menos tres configuraciones de soporte y frecuencia para cada género, edad y tipo de producto.
- ¿Qué diferencias hay entre usar soporte y frecuencia? Respalda la respuesta con datos 
- ¿Qué tipo reglas desaparecen según la configuración y categorías (género, edad y tipo de producto)
consideradas? ¿Por qué? 
- Para una de las configuraciones, interpreta algunas de las reglas que te hayan resultado interesantes
obtenidas usando la clase association_rules y un par de configuraciones cambiando la variable metric y
min_threshold. Justifica los resultados. 
- Dadas las mejores configuraciones. ¿Existen reglas o patrones que se repitan? ¿Podemos generalizar de
alguna manera como se comportan los clientes? Expón las conclusiones respaldadas con datos 

## Comprensión de los datos

La base de datos que utilizaremos consta de 233599 observaciones. Teniendo en ceunta que el trabajo que vamos a realizar, utilizanod el algoritmo apriori, queremos sacar regalas y con todos estos datos, lo que nos hace pensar es que vamos a encontrar muchísimas reglas a partir de estos datos.

La clave está en cuales son las que más nos interesan, basandonos en su confianza, y una vez hallado esto debemos tomar decisiones para que usando estas reglas vendamos el mayor número de producto, para aumentar nuestras gananzas.

Existen 11 tipos de atributos distintos: 

- User_ID: número para identificar al usuario.
- Product_ID: número para identificar el producto.
- Gender: Género del comprador.	
- Age: Edad del comprador.
- Occupation: A lo que se dedica el comprador, ya pasasdo a categórica.
- City_Category: Categoría de la ciudad donde reside el user.
- Stay_In_Current_City_Years: El tiempo que lleva en la ciudad.
- Marital_Status: Estado civil.
- Product_Category_1: La categoría principal del producto.
- Product_Category_2: Siguente categoría.
- Product_Category_3: Última categoría del producto.



##Preparación de los datos

In [9]:
usernan = df['User_ID'].isna().sum()
print(f"La cantidad de NaN que existen en la variable User_ID es:{usernan}")
productnan = df['Product_ID'].isna().sum()
print(f"La cantidad de NaN que existen en la variable Product_ID es:{productnan}")
gendernan = df['Gender'].isna().sum()
print(f"La cantidad de NaN que existen en la variable Gender es:{gendernan}")
agenan = df['Age'].isna().sum()
print(f"La cantidad de NaN que existen en la variable Age es:{agenan}")
ocupationan = df['Occupation'].isna().sum()
print(f"La cantidad de NaN que existen en la variable Occupation es:{ocupationan}")
City_Categorynan = df['City_Category'].isna().sum()
print(f"La cantidad de NaN que existen en la variable City_Category es:{City_Categorynan}")
Stay_In_Current_City_Yearsnan = df['Stay_In_Current_City_Years'].isna().sum()
print(f"La cantidad de NaN que existen en la variable Stay_In_Current_City_Years es:{Stay_In_Current_City_Yearsnan}")
Marital_Statusnan = df['Marital_Status'].isna().sum()
print(f"La cantidad de NaN que existen en la variable Marital_Status es:{Marital_Statusnan}")
Product_Category_1nan = df['Product_Category_1'].isna().sum()
print(f"La cantidad de NaN que existen en la variable Product_Category_1 es:{Product_Category_1nan}")
Product_Category_2nan = df['Product_Category_2'].isna().sum()
print(f"La cantidad de NaN que existen en la variable Product_Category_2 es:{Product_Category_2nan}")
Product_Category_3nan = df['Product_Category_3'].isna().sum()
print(f"La cantidad de NaN que existen en la variable Product_Category_3 es:{Product_Category_3nan}")

La cantidad de NaN que existen en la variable User_ID es:0
La cantidad de NaN que existen en la variable Product_ID es:0
La cantidad de NaN que existen en la variable Gender es:0
La cantidad de NaN que existen en la variable Age es:0
La cantidad de NaN que existen en la variable Occupation es:0
La cantidad de NaN que existen en la variable City_Category es:0
La cantidad de NaN que existen en la variable Stay_In_Current_City_Years es:0
La cantidad de NaN que existen en la variable Marital_Status es:0
La cantidad de NaN que existen en la variable Product_Category_1 es:0
La cantidad de NaN que existen en la variable Product_Category_2 es:72344
La cantidad de NaN que existen en la variable Product_Category_3 es:162562


##Modelado

In [70]:
te = TransactionEncoder()
te_ary = te.fit(data1).transform(data1)
data1 = pd.DataFrame(te_ary, columns=te.columns_)
data1

D      I      P      _      c      d      o      r      t      u
0        True   True   True   True   True   True   True   True   True   True
1       False  False  False  False  False  False  False  False  False  False
2       False  False  False  False  False  False  False  False  False  False
3       False  False  False  False  False  False  False  False  False  False
4       False  False  False  False  False  False  False  False  False  False
...       ...    ...    ...    ...    ...    ...    ...    ...    ...    ...
233594  False  False  False  False  False  False  False  False  False  False
233595  False  False  False  False  False  False  False  False  False  False
233596  False  False  False  False  False  False  False  False  False  False
233597  False  False  False  False  False  False  False  False  False  False
233598  False  False  False  False  False  False  False  False  False  False

[233599 rows x 10 columns]

In [7]:
pilar

P00000142  P00000242  P00000342  P00000442  P00000542  P00000642  \
0               0          0          0          0          0          0   
1               0          0          0          0          0          0   
2               0          0          0          0          0          0   
3               0          0          0          0          0          0   
4               0          0          0          0          0          0   
...           ...        ...        ...        ...        ...        ...   
233594          0          0          0          0          0          0   
233595          0          0          0          0          0          0   
233596          0          0          0          0          0          0   
233597          0          0          0          0          0          0   
233598          0          0          0          0          0          0   

        P00000742  P00000842  P00000942  P00001042  ...  P0099042  P0099142  \
0               0          0          0          0  ...         0         0   
1               0          0          0          0  ...         0         0   
2               0          0          0          0  ...         0         0   
3               0          0          0          0  ...         0         0   
4               0          0          0          0  ...         0         0   
...           ...        ...        ...        ...  ...       ...       ...   
233594          0          0          0          0  ...         0         0   
233595          0          0          0          0  ...         0         0   
233596          0          0          0          0  ...         0         0   
233597          0          0          0          0  ...         0         0   
233598          0          0          0          0  ...         0         0   

        P0099242  P0099342  P0099442  P0099542  P0099642  P0099742  P0099842  \
0              0         0         0         0         0         0         0   
1              0         0         0         0         0         0         0   
2              0         0         0         0         0         0         0   
3              0         0         0         0         0         0         0   
4              0         0         0         0         0         0         0   
...          ...       ...       ...       ...       ...       ...       ...   
233594         0         0         0         0         0         0         0   
233595         0         0         0         0         0         0         0   
233596         0         0         0         0         0         0         0   
233597         0         0         0         0         0         0         0   
233598         0         0         0         0         0         0         0   

        P0099942  
0              0  
1              0  
2              0  
3              0  
4              0  
...          ...  
233594         0  
233595         0  
233596         0  
233597         0  
233598         0  

[233599 rows x 3491 columns]

In [9]:
apriori(pilar, min_support=0.6, use_colnames=False)

Empty DataFrame
Columns: [support, itemsets]
Index: []

, use_colnames=True

In [65]:
frequent_itemsets = apriori(pilar, min_support=0.6, use_colnames=True)
frequent_itemsets['length'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))
frequent_itemsets

Empty DataFrame
Columns: [support, itemsets, length]
Index: []

In [13]:
frequent_itemsets = frequent_patterns(df['Product_ID'].apply(pd.Series).fillna(''), min_support=0.4, use_colnames=True)
print(frequent_itemsets)

TypeError: ignored

In [32]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 233599 entries, 0 to 233598
Data columns (total 3 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Product_Category_1  233599 non-null  int64  
 1   Product_Category_2  161255 non-null  float64
 2   Product_Category_3  71037 non-null   float64
dtypes: float64(2), int64(1)
memory usage: 5.3 MB


In [38]:
#data1=np.array(data1)
#(filas,columnas)=shape.data1

for i in range(233597): 
  for j in range(len(3)): 
    if data1[i][j]!= True or data1[i][j]!= False:
      print(data1[i][j])


TypeError: ignored

ESTO NO FUNCIONA 

PROBAMOS OTRO MÉTODO

In [56]:
frequent_itemsets_fr = apriori(pilar, min_support = 0.06,
  use_colnames = True)
frequent_itemsets_fr.sort_values('support', ascending = False).head()

Empty DataFrame
Columns: [support, itemsets]
Index: []

Este error da porque frequent_itemsets_fr está vacío

COPIO CÓDIGO QUE ES INTERESANTE PARA FUTURO


In [57]:
frequent_itemsets = apriori(pilar, min_support=0.6, use_colnames=True)
frequent_itemsets['length'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))
frequent_itemsets1=frequent_itemsets[frequent_itemsets['length'] == 1]

In [22]:
frequent_itemsets[ (frequent_itemsets['length'] >1 ) & #esto sería para k=2
                   (frequent_itemsets['support'] >= 0.8) ]

Empty DataFrame
Columns: [support, itemsets, length]
Index: []

Para sacar las reglas de asocuación

In [ ]:
a_rules = association_rules(frequent_itemsets_fr, metric = "lift",
  min_threshold = 1)
a_rules.sort_values('lift',ascending = False)

##Evaluación

##Despliegue